pkg/devfile/parser/reader.go

```diff
@@ -0,0 +1,107 @@
+package parser
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/devfile/library/pkg/testingutil/filesystem"
+	"github.com/devfile/library/pkg/util"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v3"
+	k8yaml "sigs.k8s.io/yaml"
+
+	routev1 "github.com/openshift/api/route/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// YamlSrc specifies the src of the yaml in either Path, URL or Data format
+type YamlSrc struct {
+	// Path is a relative or absolute yaml path.
+	Path string
+	// URL is the URL address of the specific yaml.
+	URL string
+	// Data is the yaml content in []byte format.
+	Data []byte
+}
+
+// ReadKubernetesYaml reads a yaml Kubernetes file from either the Path, URL or Data provided.
+// It returns Deployments, Services, Routes resources as the primary Kubernetes resources.
+// Other Kubernetes resources are returned as []byte type. Consumers interested in other Kubernetes resources
+// are expected to Unmarshal it to the struct of the respective resource.
+func ReadKubernetesYaml(src YamlSrc, fs filesystem.Filesystem) ([]appsv1.Deployment, []corev1.Service, []routev1.Route, [][]byte, error) {
+
+	var data []byte
+	var err error
+
+	if src.URL != "" {
+		data, err = util.DownloadFileInMemory(src.URL)
+		if err != nil {
+			return nil, nil, nil, nil, errors.Wrapf(err, "failed to download file %q", src.URL)
+		}
+	} else if src.Path != "" {
+		absPath, err := util.GetAbsPath(src.Path)
```

Comment:
probably a dumb question... is the `src.Path` you passing in here a relative path to devfile.yaml?
so the `GetAbsPath` is calsulating from the user's current path... if the devfile.yaml is under `/user/devfilefolder/devfile.yaml` and user is running parser from `/user/`.  will the `GetAbsPath()` try to look the relative path against `/user/` instead of `/user/devfilefolder/`?

Reply:
and.. what if the devfile.yaml we are parsing is from a URL? it cannot be resolved by `GetAbsPath`

Reply:
For URL there is `src.URL`; let me take a look at the AbsPath case

Reply:
Updated, not getting the abs path. Leaving it up to the consumer to handle the path. Utility func job is to just read from the path/url/data provided